### ISSUE TITLE
[Consensus] Possible signature validation bug

### ISSUE THREAD

**Issue by zeptin** [stratisproject/StratisBitcoinFullNode#2247]
While working on #2151 I found the following apparent discrepancy between stratisX and SBFN:

stratisX:
```
bool IsDERSignature(const valtype &vchSig, bool haveHashType) {
    // See https://bitcointalk.org/index.php?topic=8392.msg127623#msg127623
    // A canonical signature exists of: <30> <total len> <02> <len R> <R> <02> <len S> <S> <hashtype>
    // Where R and S are not negative (their first byte has its highest bit not set), and not
    // excessively padded (do not start with a 0 byte, unless an otherwise negative number follows,
    // in which case a single 0 byte is necessary and even required).
    if (vchSig.size() < 9)
        return error("Non-canonical signature: too short");
    if (vchSig.size() > 73)
        return error("Non-canonical signature: too long");
    if (vchSig[0] != 0x30)
        return error("Non-canonical signature: wrong type");
    if (vchSig[1] != vchSig.size() - (haveHashType ? 3 : 2))
        return error("Non-canonical signature: wrong length marker");
```

SBFN:
```
        public static bool IsValidSignatureEncoding(byte[] sig)
        {
            // Format: 0x30 [total-length] 0x02 [R-length] [R] 0x02 [S-length] [S] [sighash]
            // * total-length: 1-byte length descriptor of everything that follows,
            //   excluding the sighash byte.
            // * R-length: 1-byte length descriptor of the R value that follows.
            // * R: arbitrary-length big-endian encoded R value. It must use the shortest
            //   possible encoding for a positive integers (which means no null bytes at
            //   the start, except a single one when the next byte has its highest bit set).
            // * S-length: 1-byte length descriptor of the S value that follows.
            // * S: arbitrary-length big-endian encoded S value. The same rules apply.
            // * sighash: 1-byte value indicating what data is hashed (not part of the DER
            //   signature)

            int signLen = sig.Length;

            // Minimum and maximum size constraints.
            if(signLen < 9 || signLen > 73)
                return false;

            // A signature is of type 0x30 (compound).
            if(sig[0] != 0x30)
                return false;

            // Make sure the length covers the entire signature.
            if(sig[1] != signLen - 3)
                return false;
```

In stratisX the `ProcessBlock` method ultimately calls `IsDERSignature` with the boolean set to `false`, implying that the `hashType` is missing from the signature & that the constant `2` should be used to subtract from the length. In SBFN we do not have the bool, but by using `3` we are assuming the `hashType` is in fact affixed to the signature.

This does not appear to always be the case, e.g. while processing block 35 on Stratis testnet the following block signature was extracted:

`BlockSignature` = `3045022100ee488514f8f8dbad8176b4a314d47e22b4be30a335ac4032225b4d85cb83985a02206bcfd07da2fd1622fda056af2881fdfa1a2e27ad9dbecd35ab73ea91a31e8bf9`

```
            // 71 bytes in total (>= 9 && <= 73) [signLen = 71]

            // 0x30 -> type is correct
            // 0x45 = total len = 69
            // 0x02
            // 0x21 = len R = 33
            // 00ee488514f8f8dbad8176b4a314d47e22b4be30a335ac4032225b4d85cb83985a = R
            // 0x02
            // 0x20 = len S = 32
            // 6bcfd07da2fd1622fda056af2881fdfa1a2e27ad9dbecd35ab73ea91a31e8bf9
            // -> no hashType
```

Therefore this block fails the tests in `IsValidSignatureEncoding`. These tests are not currently used in the codebase but will be introduced at some point via #2151.

**Comment by zeptin** [comment]
Note that block 35 seems to be the first PoS block on testnet.

**Comment by dangershony** [comment]
>This does not appear to always be the case, e.g. while processing block 35 on Stratis testnet the following block signature was extracted:
>Note that block 35 seems to be the first PoS block on testnet.

So it seems POS blocks do not append the sighash?
